common/record/_internal/recordNodesIndex.js

```diff
@@ -74,60 +91,84 @@ const addNodes = nodes =>
+  //node def index
+  addOrDeleteElementInArrayInPath([keys.nodesByDef, Node.getNodeDefUuid(node)], Node.getUuid(node)),
+  //code dependent index
+  _addNodeToCodeDependentsIndex(node),
+  //dirty index
+  addOrDeleteElementInArrayInPath([keys.nodesDirty], Node.getUuid(node), Node.isDirty(node))
+)
 
-const _addNodeToParentIndex = node => record => R.pipe(
-  getNodeUuidsByParentAndDef(Node.getParentUuid(node), Node.getNodeDefUuid(node)),
-  R.ifElse(
-    R.includes(Node.getUuid(node)),
-    R.identity,
-    R.append(Node.getUuid(node))
-  ),
-  arr => R.assocPath([keys.nodesByParentAndDef, Node.getParentUuid(node), Node.getNodeDefUuid(node)], arr, record)
-)(record)
+const _addNodeToCodeDependentsIndex = node => record =>
+  R.reduce(
+    (recordAcc, ancestorCodeAttributeUuid) =>
+      addOrDeleteElementInArrayInPath([keys.nodeCodeDependents, ancestorCodeAttributeUuid], Node.getUuid(node))(recordAcc),
+    record,
+    Node.getCodeAttributeHierarchy(node)
+  )
 
 // ===== DELETE
 /**
  * Removed the specified node from the cache
  */
 const removeNode = node => R.pipe(
-  _removeNodeFromParentIndex(node),
-  _removeNodeFromNodeDefIndex(node)
+  R.ifElse(
+    R.always(Node.isRoot(node)),
+    R.dissoc(keys.nodeRootUuid),
+    R.identity
+  ),
+  //parent index
+  addOrDeleteElementInArrayInPath([keys.nodesByParentAndDef, Node.getParentUuid(node), Node.getNodeDefUuid(node)], Node.getUuid(node), false),
+  //node def index
+  addOrDeleteElementInArrayInPath([keys.nodesByDef, Node.getNodeDefUuid(node)], Node.getUuid(node), false),
+  //code dependent index
+  _removeNodeFromCodeDependentsIndex(node),
+  //dirty index
+  addOrDeleteElementInArrayInPath([keys.nodesDirty], Node.getUuid(node), false)
 )
 
-const _removeNodeFromParentIndex = node => record => R.pipe(
-  getNodeUuidsByParentAndDef(Node.getParentUuid(node), Node.getNodeDefUuid(node)),
-  R.without([Node.getUuid(node)]),
-  arr => R.assocPath([keys.nodesByParentAndDef, Node.getParentUuid(node), Node.getNodeDefUuid(node)], arr)(record)
-)(record)
-
-const _removeNodeFromNodeDefIndex = node => record => R.pipe(
-  getNodeUuidsByDef(Node.getNodeDefUuid(node)),
-  R.without([Node.getUuid(node)]),
-  arr => R.assocPath([keys.nodesByDef, Node.getNodeDefUuid(node)], arr)(record)
-)(record)
+const _removeNodeFromCodeDependentsIndex = node => record => R.pipe(
+  Node.getCodeAttributeHierarchy,
+  R.reduce(
+    (recordAcc, ancestorCodeAttributeUuid) =>
+      addOrDeleteElementInArrayInPath([keys.nodeCodeDependents, ancestorCodeAttributeUuid], Node.getUuid(node), false)(recordAcc),
+    record,
+  ),
+  R.dissocPath([keys.nodeCodeDependents, Node.getUuid(node)])
+)(node)
 
+const addOrDeleteElementInArrayInPath = (path, value, add = true) => obj => R.pipe(
+  R.pathOr([], path),
+  R.ifElse(
+    R.always(add),
+    R.ifElse(
+      R.includes(value),
+      R.identity,
+      R.append(value)
+    ),
+    R.without(value)
+  ),
+  arr => R.assocPath(path, arr, obj)
+)(obj)
 
```

Comment:
It's not very clear to me have a method that does both. Can we have two? Like ```_assocToIndexPath``` and ```_dissocToIndexPath```?

Reply:
ok